### PR TITLE
Multiline %r

### LINKED ISSUE #1020: [PROPOSAL] Allow %r{} syntax for multiline regexps
IMHO %r{} looks better for multi=line regexps:

``` ruby
RE = %r{
  ^
    (...)
    (...)
  $
}x

# vs

RE = /
  ^
    (...)
    (...)
  $
/x
```

### PR DESCRIPTION
**Pull request by bquorning**
Fixes #1020.

Following the discussion on #1655 (see https://github.com/bbatsov/rubocop/pull/1655#issuecomment-74412384 and https://github.com/bbatsov/rubocop/pull/1655#discussion_r26288686), and https://github.com/bbatsov/ruby-style-guide/commit/bdde52a668da9fec99ba4e33f893bce91d18494f, this PR removes the `MaxSlashes` option from the `RegexpLiteral` cop, instead adding the `EnforcedStyle` option (either `slashes`, `percent_r` or `mixed`) and the `AllowInnerSlashes` option.

The implementation is very similar to how `CommandLiteral` was implemented in #1655.

The former implementation had some `auto_gen_config` logic that I have completely removed. I *think* all it did was keep a counter to set the `MaxSlashes` value in the generated config file.

### CODE REVIEW

File: config/enabled.yml [hunk 1]
```diff
@@ -499,11 +499,7 @@ Style/RedundantSelf:
   Enabled: true
 
 Style/RegexpLiteral:
-  Description: >-
-                 Use %r for regular expressions matching more than
-                 `MaxSlashes` '/' characters.
-                 Use %r only for regular expressions matching more than
-                 `MaxSlashes` '/' character.
+  Description: 'Use / or %r around regexp literals.'
```
Comment (bbatsov) [inline config/enabled.yml]: - around; the delimiters are part of the literals

File: config/default.yml [hunk 2]
```diff
@@ -453,10 +453,19 @@ Style/RedundantReturn:
   # When true allows code like `return x, y`.
   AllowMultipleReturnValues: false
 
+# Use / or %r around regexp literals.
 Style/RegexpLiteral:
-  # The maximum number of (escaped) slashes that a slash-delimited regexp is
-  # allowed to have. If there are more slashes, a %r regexp shall be used.
-  MaxSlashes: 1
+  EnforcedStyle: slashes
+  # slashes: Always use slashes.
```
Comment (bbatsov) [inline config/default.yml]: `%r` should always be allowed for a literal containing `/`.

File: config/enabled.yml (hunk 1 above)
Comment (bquorning) [inline config/enabled.yml]: Maybe `'Use / or %r around regular expressions.'`?

**Comment by bquorning** [comment]
It looks like this was the only cop using `auto_gen_config`. Can I just remove the failing spec?

**Comment by bbatsov** [comment]
Guess so.

On 18 March 2015 at 11:47, Benjamin Quorning <notifications@github.com>
wrote:

> It looks like this was the only cop using auto_gen_config. Can I just
> remove the failing spec?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bbatsov/rubocop/pull/1721#issuecomment-82855807>.
>



-- 
Best Regards,
Bozhidar Batsov

http://www.batsov.com

**Comment by bquorning** [comment]
I managed to change the failing cli_spec example, so it still tests the usefulness of the auto-generated config file.

Hopefully, this fixes the test coverage report.

**Comment by bquorning** [comment]
So, the coverage dropped because I removed the line `RuboCop::Cop::Style::RegexpLiteral.slash_count = 0` from cli_spec. That line caused an error to happen in https://github.com/bbatsov/rubocop/blob/241a70837f5d9d0db9bcef996318040293cb8888/lib/rubocop/cop/style/regexp_literal.rb#L47 where slash_count is expected to be a hash.

I think that writing a new spec for `Team#process_commissioner_errors` and `Team#process_commissioner_errors` is for another pull request.

**Comment by bbatsov** [comment]
:+1: Good work!

> I think that writing a new spec for Team#process_commissioner_errors and Team#process_commissioner_errors is for another pull request.

Makes sense.

**Comment by bquorning** [comment]
> :+1: Good work!

Thanks

**Pull request merged**